src/nginx_manager.py

```diff
@@ -0,0 +1,260 @@
+# Copyright 2024 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Manage nginx instance."""
+
+import logging
+import shutil
+import uuid
+from pathlib import Path
+
+import nginx
+
+from errors import (
+    NginxConfigurationAggregateError,
+    NginxConfigurationError,
+    NginxFileError,
+    NginxSetupError,
+    NginxStopError,
+)
+from state import HostConfig, NginxConfig
+from utilities import execute_command
+
+logger = logging.getLogger(__name__)
+
+NGINX_SITES_ENABLED_PATH = Path("/etc/nginx/sites-enabled")
+NGINX_SITES_AVAILABLE_PATH = Path("/etc/nginx/sites-available")
+NGINX_LOG_PATH = Path("/var/log/nginx")
+
+
+# Unit test is not valuable as the module is closely coupled with nginx.
+# This should be tested with integration tests.
+
+
+def initialize() -> None:  # pragma: no cover
+    """Initialize the nginx server.
+
+    Raises:
+        NginxSetupError: Failure to set up nginx.
+    """
+    logger.info("Installing and enabling nginx")
+    # The install, systemctl enable, and systemctl start are idempotent.
+    return_code, _, stderr = execute_command(["sudo", "apt", "install", "nginx", "-yq"])
+    if return_code != 0:
+        raise NginxSetupError(f"Failed to install nginx: {stderr}")
+
+    logger.info("Clean up default configuration files")
+    _reset_sites_config_files()
+    return_code, _, stderr = execute_command(["sudo", "systemctl", "enable", "nginx"])
+    if return_code != 0:
+        raise NginxSetupError(f"Failed to enable nginx: {stderr}")
+    return_code, _, stderr = execute_command(["sudo", "systemctl", "start", "nginx"])
+    if return_code != 0:
+        raise NginxSetupError(f"Failed to start nginx: {stderr}")
+
+
+def load_config() -> None:  # pragma: no cover
+    """Load nginx configurations."""
+    if ready_check():
+        logger.info("Loading nginx configuration files")
+        # This is reload the configuration files without interrupting service.
+        execute_command(["sudo", "nginx", "-s", "reload"])
+        return
+    logger.info("Restarting nginx to load the configuration files.")
```

Comment:
nitpick: I would leave an empty line here to indicate that these are different code blocks